### PR TITLE
Sending project format warnings in DTH and update DiagnosticsMessage

### PR DESCRIPTION
**Pull request by troydai**
@BillHiebert @davidfowl 

For issue https://github.com/aspnet/DNX/issues/1330

### CODE REVIEW

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs [hunk 1]
```diff
@@ -372,9 +372,11 @@ private bool ResolveDependencies()
 
                 ProjectSearchPaths = state.ProjectSearchPaths,
 
-                GlobalJsonPath = state.GlobalJsonPath
+                GlobalJsonPath = state.GlobalJsonPath,
```
Comment (muratg) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Why did you need to add the comma?

Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Shouldn't. It's a left-over in the intermediate change. Will revert.

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs [hunk 2]
```diff
@@ -433,11 +435,9 @@ private bool PerformCompilation()
...
                     projectCompilationChanged = UpdateProjectCompilation(project, out compilation);
 
-                    project.Diagnostics = new DiagnosticsMessage
-                    {
-                        Framework = project.Sources.Framework,
-                        Diagnostics = compilation.Diagnostics
-                    };
+                    project.Diagnostics = new DiagnosticsMessage(
+                        compilation.Diagnostics,
+                        project.Sources.Framework);
```
Comment (muratg) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Is his change changing anything semantically? Is it required?

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs [hunk 3]
```diff
@@ -477,11 +477,9 @@ private bool PerformCompilation()
...
                     {
-                        project.Diagnostics = new DiagnosticsMessage
-                        {
-                            Framework = project.Sources.Framework,
-                            Diagnostics = compilation.Diagnostics,
-                        };
+                        project.Diagnostics = new DiagnosticsMessage(
+                            compilation.Diagnostics,
+                            project.Sources.Framework);
                     }
```
Comment (muratg) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Ditto

File: src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs [hunk 4]
```diff
@@ -10,31 +10,65 @@ namespace Microsoft.Framework.DesignTimeHost.Models.OutgoingMessages
...
+            Framework = frameworkData;
+            CompilationDiagnostics = compilationMessages;
         }
 
-        public override int GetHashCode()
+        public DiagnosticsMessage(IEnumerable<DiagnosticsInfo> errors, IEnumerable<DiagnosticsInfo> warnings, FrameworkData framework)
         {
-            // These objects are currently POCOs and we're overriding equals
-            // so that things like Enumerable.SequenceEqual just work.
-            return base.GetHashCode();
```
Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs]: I'll add the override of equals back.

File: src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs [hunk 5]
```diff
@@ -10,31 +10,65 @@ namespace Microsoft.Framework.DesignTimeHost.Models.OutgoingMessages
...
-        public FrameworkData Framework { get; set; }
-
-        [JsonIgnore]
-        public IList<ICompilationMessage> Diagnostics { get; set; }
-
-        public IEnumerable<string> Errors => Diagnostics.Where(d => d.Severity == CompilationMessageSeverity.Error)
-                                                        .Select(d => d.FormattedMessage);
+        private object error;
+        private IEnumerable<DiagnosticsInfo> warnings;
+        private object framework;
```
Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs]: These will be removed.

File: src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs [hunk 6]
```diff
@@ -10,31 +10,65 @@ namespace Microsoft.Framework.DesignTimeHost.Models.OutgoingMessages
...
+        public IList<ICompilationMessage> CompilationDiagnostics { get; }
+
+        public FrameworkData Framework { get; }
+
+        public IEnumerable<DiagnosticsInfo> Errors { get; }
+
+        public IEnumerable<DiagnosticsInfo> Warnings { get; }
+    }
+
+    public class DiagnosticsInfo
```
Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs]: This will be move to its own file.

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs (hunk 2 above)
Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: The `DiagnosticsMessage` now also carries diagnostics information other than _compilation diagnostics_. The original interface doesn't distinguish two initialization approaches well enough.

**Comment by troydai** [comment]
@muratg @davidfowl @anurse @victorhurdugaci

**Comment by troydai** [comment]
Ping @BillHiebert @davidfowl @muratg

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs [hunk 7]
```diff
@@ -631,6 +636,33 @@ private void SendOutgoingMessages()
...
+                    Path = warning.Path,
+                    FormattedMessage = warning.Message
+                }),
+                errors: null,
+                framework: null);
+
+            _initializedContext.Transmit(new Message
+            {
+                ContextId = Id,
+                MessageType = "Diagnostics",
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Const?

Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Maybe we should. But for now none of the `MessageType` value are given in const, but asall hard coded.

File: src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs [hunk 8]
```diff
@@ -10,30 +10,91 @@ namespace Microsoft.Framework.DesignTimeHost.Models.OutgoingMessages
...
+                 other.Line == Line && 
+                 other.Column == Column && 
+                 other.Message == Message && 
+                 other.FormattedMessage == FormattedMessage;
         }
 
         public override int GetHashCode()
         {
-            // These objects are currently POCOs and we're overriding equals
-            // so that things like Enumerable.SequenceEqual just work.
```
Comment (muratg) [inline src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs]: Is this comment no longer valid?  If so, do we need this override?

**Comment by muratg** [comment]
Looks good to me, though I'd like to see @BillHiebert have a look as well.  :shipit:

File: src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs (hunk 8 above)
Comment (troydai) [inline src/Microsoft.Framework.DesignTimeHost/Models/OutgoingMessages/DiagnosticsMessage.cs]: Yeah, I thought it's unnecessary but it's needed actually. If `Equals` is overridden, the `GetHashCode` must be overriden. [CA2218(https://msdn.microsoft.com/en-us/library/ms182358.aspx)]

**Comment by troydai** [comment]
Take this off. Need more work.
